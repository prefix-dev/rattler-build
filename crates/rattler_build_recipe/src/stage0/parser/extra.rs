//! Parser for the Extra section

use marked_yaml::Node as MarkedNode;

use crate::{
    error::{ParseError, ParseResult},
    stage0::{extra::Extra, parser::helpers::get_span, parser::list::parse_conditional_list},
};

/// Parse an Extra section from YAML
///
/// The Extra section contains metadata like recipe maintainers.
///
/// Example YAML:
/// ```yaml
/// extra:
///   recipe-maintainers:
///     - alice
///     - bob
/// ```
pub fn parse_extra(yaml: &MarkedNode) -> ParseResult<Extra> {
    let mapping = yaml.as_mapping().ok_or_else(|| {
        ParseError::expected_type("mapping", "non-mapping", get_span(yaml))
            .with_message("Extra section must be a mapping")
    })?;

    let mut extra = Extra::default();

    // Parse recipe-maintainers field
    if let Some(maintainers) = mapping.get("recipe-maintainers") {
        extra.recipe_maintainers = parse_conditional_list(maintainers)?;
    }

    // Check for unknown fields
    for (key, _) in mapping.iter() {
        let key_str = key.as_str();
        if !matches!(key_str, "recipe-maintainers") {
            return Err(ParseError::invalid_value(
                "extra",
                &format!("unknown field '{}'", key_str),
                (*key.span()).into(),
            )
            .with_suggestion("valid fields are: recipe-maintainers"));
        }
    }

    Ok(extra)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn parse_yaml_extra(yaml_str: &str) -> MarkedNode {
        let wrapped = format!("extra:\n{}", yaml_str);
        let root = marked_yaml::parse_yaml(0, &wrapped).expect("Failed to parse test YAML");
        let mapping = root.as_mapping().expect("Expected mapping");
        mapping.get("extra").expect("Field not found").clone()
    }

    #[test]
    fn test_parse_empty_extra() {
        let yaml = parse_yaml_extra("  {}");
        let extra = parse_extra(&yaml).unwrap();
        assert!(extra.recipe_maintainers.is_empty());
    }

    #[test]
    fn test_parse_extra_with_maintainers() {
        let yaml_str = r#"
  recipe-maintainers:
    - alice
    - bob
    - charlie"#;
        let yaml = parse_yaml_extra(yaml_str);
        let extra = parse_extra(&yaml).unwrap();
        assert_eq!(extra.recipe_maintainers.len(), 3);
    }

    #[test]
    fn test_parse_extra_with_template_maintainers() {
        let yaml_str = r#"
  recipe-maintainers:
    - '${{ maintainer_1 }}'
    - bob"#;
        let yaml = parse_yaml_extra(yaml_str);
        let extra = parse_extra(&yaml).unwrap();
        assert_eq!(extra.recipe_maintainers.len(), 2);

        // Verify template variable extraction
        let vars = extra.used_variables();
        assert_eq!(vars, vec!["maintainer_1"]);
    }

    #[test]
    fn test_parse_extra_unknown_field() {
        let yaml_str = r#"
  recipe-maintainers:
    - alice
  unknown-field: value"#;
        let yaml = parse_yaml_extra(yaml_str);
        let result = parse_extra(&yaml);

        assert!(result.is_err());
        let err = result.unwrap_err();
        let err_string = err.to_string();
        assert!(err_string.contains("unknown field"));
    }

    #[test]
    fn test_parse_extra_not_mapping() {
        let wrapped = "extra: not a mapping";
        let root = marked_yaml::parse_yaml(0, wrapped).expect("Failed to parse test YAML");
        let mapping = root.as_mapping().expect("Expected mapping");
        let yaml = mapping.get("extra").expect("Field not found");

        let result = parse_extra(yaml);
        assert!(result.is_err());
        let err = result.unwrap_err();
        let err_string = err.to_string();
        assert!(err_string.contains("mapping") || err_string.contains("expected"));
    }
}
