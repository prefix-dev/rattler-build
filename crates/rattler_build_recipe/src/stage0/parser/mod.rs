//! Parser for converting YAML to stage0 recipe structures
//!
//! This module provides parsing functionality that converts YAML (via marked_yaml)
//! into stage0 recipe types. All parsing preserves span information for excellent
//! error messages.

mod about;
mod build;
mod extra;
mod helpers;
mod list;
mod output_parser;
mod package;
mod requirements;
mod source;
mod test_parser;
mod value;

#[cfg(test)]
mod error_tests;
#[cfg(test)]
mod recipe_tests;
#[cfg(test)]
mod snapshot_tests;
#[cfg(test)]
mod unit_tests;

use marked_yaml::Node as MarkedNode;

use crate::{
    error::{ErrorKind, ParseError, ParseResult},
    span::Span,
};

// Re-export parsing functions
pub use about::parse_about;
pub use build::parse_build;
pub use extra::parse_extra;
pub use list::parse_conditional_list;
pub use output_parser::parse_multi_output_recipe;
pub use package::parse_package;
pub use requirements::parse_requirements;
pub use source::parse_source;
pub use test_parser::parse_tests;
pub use value::{parse_value, parse_value_with_name};

// Re-export helpers within crate only
pub(crate) use helpers::get_span;

/// Parse a recipe (single or multi-output) from YAML source string
///
/// This function automatically detects whether the recipe is single-output or multi-output
/// and returns the appropriate Recipe variant.
pub fn parse_recipe_or_multi_from_source(source: &str) -> ParseResult<crate::stage0::Recipe> {
    let yaml = marked_yaml::parse_yaml(0, source).map_err(|e| {
        ParseError::new(ErrorKind::YamlError, Span::unknown())
            .with_message(format!("Failed to parse YAML: {}", e))
    })?;

    parse_recipe_or_multi(&yaml)
}

/// Parse a complete stage0 recipe from YAML source string
///
/// Note: This function returns a SingleOutputRecipe for backwards compatibility.
/// For multi-output recipe support, use `parse_recipe_or_multi_from_source()`.
pub fn parse_recipe_from_source(source: &str) -> ParseResult<crate::stage0::Stage0Recipe> {
    let yaml = marked_yaml::parse_yaml(0, source).map_err(|e| {
        ParseError::new(ErrorKind::YamlError, Span::unknown())
            .with_message(format!("Failed to parse YAML: {}", e))
    })?;

    parse_recipe(&yaml)
}

/// Parse a recipe (single or multi-output) from YAML
///
/// This function automatically detects whether the recipe is single-output or multi-output:
/// - If the recipe has an "outputs" key, it's parsed as a multi-output recipe
/// - Otherwise, it's parsed as a single-output recipe
///
/// Multi-output recipes use a "recipe" section instead of "package" at the top level.
pub fn parse_recipe_or_multi(yaml: &MarkedNode) -> ParseResult<crate::stage0::Recipe> {
    let mapping = yaml.as_mapping().ok_or_else(|| {
        ParseError::expected_type("mapping", "non-mapping", helpers::get_span(yaml))
            .with_message("Recipe must be a mapping")
    })?;

    // Detect multi-output by presence of "outputs" key
    if mapping.get("outputs").is_some() {
        // Multi-output recipe
        let multi = parse_multi_output_recipe(mapping)?;
        Ok(crate::stage0::Recipe::MultiOutput(Box::new(multi)))
    } else {
        // Single-output recipe
        let single = parse_single_output_recipe(yaml)?;
        Ok(crate::stage0::Recipe::SingleOutput(Box::new(single)))
    }
}

/// Parse a complete stage0 recipe from YAML
///
/// Note: This function parses single-output recipes only.
/// For multi-output recipe support, use `parse_recipe_or_multi()`.
///
/// The recipe must be a mapping with at minimum a `package` section.
/// Other sections (about, requirements, extra) are optional.
///
/// Example YAML:
/// ```yaml
/// package:
///   name: my-package
///   version: 1.0.0
/// about:
///   license: MIT
///   summary: A test package
/// requirements:
///   build:
///     - gcc
///   run:
///     - python
/// extra:
///   recipe-maintainers:
///     - alice
/// ```
pub fn parse_recipe(yaml: &MarkedNode) -> ParseResult<crate::stage0::Stage0Recipe> {
    parse_single_output_recipe(yaml)
}

/// Parse a single-output recipe from YAML
///
/// Internal function used by both parse_recipe and parse_recipe_or_multi.
fn parse_single_output_recipe(yaml: &MarkedNode) -> ParseResult<crate::stage0::SingleOutputRecipe> {
    let mapping = yaml.as_mapping().ok_or_else(|| {
        ParseError::expected_type("mapping", "non-mapping", helpers::get_span(yaml))
            .with_message("Recipe must be a mapping")
    })?;

    // Parse optional schema_version
    let schema_version = if let Some(schema_node) = mapping.get("schema_version") {
        let scalar = schema_node.as_scalar().ok_or_else(|| {
            ParseError::expected_type("scalar", "non-scalar", helpers::get_span(schema_node))
                .with_message("schema_version must be an integer")
        })?;
        let version_str = scalar.as_str();
        let version: u32 = version_str.parse().map_err(|_| {
            ParseError::invalid_value(
                "schema_version",
                "not a valid integer",
                (*scalar.span()).into(),
            )
        })?;

        // Only version 1 is supported
        if version != 1 {
            return Err(ParseError::invalid_value(
                "schema_version",
                &format!(
                    "unsupported schema version {} (only version 1 is supported)",
                    version
                ),
                (*scalar.span()).into(),
            ));
        }
        Some(version)
    } else {
        None
    };

    // Parse optional context
    let context = if let Some(context_node) = mapping.get("context") {
        parse_context(context_node)?
    } else {
        indexmap::IndexMap::new()
    };

    // Parse required package section
    let package_node = mapping
        .get("package")
        .ok_or_else(|| ParseError::missing_field("package", helpers::get_span(yaml)))?;
    let package = parse_package(package_node)?;

    // Parse optional sections (will use default if not present)
    let build = if let Some(build_node) = mapping.get("build") {
        parse_build(build_node)?
    } else {
        crate::stage0::Build::default()
    };

    let about = if let Some(about_node) = mapping.get("about") {
        parse_about(about_node)?
    } else {
        crate::stage0::About::default()
    };

    let requirements = if let Some(requirements_node) = mapping.get("requirements") {
        parse_requirements(requirements_node)?
    } else {
        crate::stage0::Requirements::default()
    };

    let extra = if let Some(extra_node) = mapping.get("extra") {
        parse_extra(extra_node)?
    } else {
        crate::stage0::Extra::default()
    };

    // Parse optional source section (can be empty)
    let source = if let Some(source_node) = mapping.get("source") {
        parse_source(source_node)?
    } else {
        Vec::new()
    };

    // Parse optional tests section (can be empty)
    let tests = if let Some(tests_node) = mapping.get("tests") {
        parse_tests(tests_node)?
    } else {
        Vec::new()
    };

    // Check for unknown top-level fields
    for (key, _) in mapping.iter() {
        let key_str = key.as_str();
        if !matches!(
            key_str,
            "package"
                | "build"
                | "about"
                | "requirements"
                | "extra"
                | "source"
                | "tests"
                | "schema_version"
                | "context"
        ) {
            return Err(ParseError::invalid_value(
                "recipe",
                &format!("unknown top-level field '{}'", key_str),
                (*key.span()).into(),
            )
            .with_suggestion("valid top-level fields are: package, build, about, requirements, extra, source, tests, schema_version, context"));
        }
    }

    Ok(crate::stage0::SingleOutputRecipe {
        schema_version,
        context,
        package,
        build,
        about,
        requirements,
        extra,
        source,
        tests,
    })
}

/// Parse the context section from YAML
///
/// Context is an order-preserving mapping of variable names to values (can be templates or concrete).
/// Order is important because later context values can reference earlier ones.
pub(crate) fn parse_context(
    yaml: &MarkedNode,
) -> ParseResult<
    indexmap::IndexMap<String, crate::stage0::types::Value<rattler_build_jinja::Variable>>,
> {
    let mapping = yaml.as_mapping().ok_or_else(|| {
        ParseError::expected_type("mapping", "non-mapping", helpers::get_span(yaml))
            .with_message("context must be a mapping")
    })?;

    let mut context = indexmap::IndexMap::new();

    for (key_node, value_node) in mapping.iter() {
        let key = key_node.as_str().to_string();

        // Check for hyphens in variable names
        if key.contains('-') {
            return Err(ParseError::invalid_value(
                "context variable name",
                "variable names cannot contain hyphens (-) as they are not valid in Jinja expressions",
                (*key_node.span()).into(),
            ));
        }

        let value = parse_context_value(value_node, &key)?;
        context.insert(key, value);
    }

    Ok(context)
}

/// Parse a context value - can be either a scalar or a list of uniform scalars
fn parse_context_value(
    yaml: &MarkedNode,
    key: &str,
) -> ParseResult<crate::stage0::types::Value<rattler_build_jinja::Variable>> {
    use rattler_build_jinja::Variable;

    // Check if it's a sequence (list)
    if let Some(sequence) = yaml.as_sequence() {
        // Parse list of uniform scalar values
        let mut variables = Vec::new();
        let mut expected_kind: Option<minijinja::value::ValueKind> = None;

        for (index, item_node) in sequence.iter().enumerate() {
            // Each item must be a scalar
            let scalar = item_node.as_scalar().ok_or_else(|| {
                ParseError::expected_type("scalar", "non-scalar", helpers::get_span(item_node))
                    .with_message(format!(
                        "context.{}[{}] must be a scalar (string, number, or boolean)",
                        key, index
                    ))
            })?;

            let s = scalar.as_str();

            // Parse the scalar value based on its type:
            // - Booleans: true/false remain as booleans
            // - Integers: whole numbers without decimals remain as integers
            // - Floats: numbers with decimals become strings (to preserve versions like "1.23")
            // - Everything else: strings
            let variable = if s == "true" || s == "false" {
                Variable::from(s.parse::<bool>().unwrap())
            } else if let Ok(int_val) = s.parse::<i64>() {
                // Only treat as integer if it doesn't contain a decimal point
                if !s.contains('.') {
                    Variable::from(int_val)
                } else {
                    Variable::from(s.to_string())
                }
            } else {
                Variable::from(s.to_string())
            };

            variables.push(variable);
        }

        // Return a concrete value containing a Variable list
        let list_variable = Variable::from(variables);
        Ok(crate::stage0::types::Value::new_concrete(
            list_variable,
            helpers::get_span(yaml),
        ))
    } else if let Some(scalar) = yaml.as_scalar() {
        // Parse scalar value
        let spanned = crate::span::SpannedString::from(scalar);
        let s = spanned.as_str();
        let span = spanned.span();

        // Check if it contains a Jinja template
        if s.contains("${{") && s.contains("}}") {
            // It's a template - we'll store it as a string template for now
            // The actual Variable will be created during evaluation
            let template = crate::stage0::types::JinjaTemplate::new(s.to_string())
                .map_err(|e| ParseError::jinja_error(e, span))?;
            Ok(crate::stage0::types::Value::new_template(template, span))
        } else {
            // Parse the scalar value based on its type:
            // - Booleans: true/false remain as booleans
            // - Integers: whole numbers without decimals remain as integers
            // - Floats: numbers with decimals become strings (to preserve versions like "1.23")
            // - Everything else: strings
            let variable = if s == "true" || s == "false" {
                Variable::from(s.parse::<bool>().unwrap())
            } else if let Ok(int_val) = s.parse::<i64>() {
                // Only treat as integer if it doesn't contain a decimal point
                if !s.contains('.') {
                    Variable::from(int_val)
                } else {
                    Variable::from(s.to_string())
                }
            } else {
                Variable::from(s.to_string())
            };

            Ok(crate::stage0::types::Value::new_concrete(variable, span))
        }
    } else {
        Err(ParseError::expected_type(
            "scalar or list",
            "non-scalar/non-list",
            helpers::get_span(yaml),
        ).with_message("context values must be scalars (strings, numbers, booleans) or lists of uniform scalars"))
    }
}

// All section parsers (parse_package, parse_about, parse_extra, parse_requirements)
// are now implemented in their respective modules and re-exported above
