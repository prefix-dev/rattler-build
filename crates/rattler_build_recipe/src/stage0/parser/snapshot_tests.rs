//! Snapshot tests for recipe parsing using insta
//!
//! These tests parse real recipe files from test-data/ and snapshot the results

use crate::stage0::parser::parse_recipe_from_source;

const TEST_DATA_DIR: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/test-data");

fn load_test_recipe(filename: &str) -> String {
    let path = format!("{}/{}", TEST_DATA_DIR, filename);
    fs_err::read_to_string(&path)
        .unwrap_or_else(|e| panic!("Failed to read test file {}: {}", path, e))
}

#[test]
fn test_minimal_recipe_snapshot() {
    let source = load_test_recipe("minimal.yaml");
    let recipe = parse_recipe_from_source(&source).expect("Failed to parse minimal recipe");
    insta::assert_debug_snapshot!(recipe);
}

#[test]
fn test_full_recipe_snapshot() {
    let source = load_test_recipe("full.yaml");
    let recipe = parse_recipe_from_source(&source).expect("Failed to parse full recipe");
    insta::assert_debug_snapshot!(recipe);
}

#[test]
fn test_templates_recipe_snapshot() {
    let source = load_test_recipe("templates.yaml");
    let recipe = parse_recipe_from_source(&source).expect("Failed to parse templates recipe");
    insta::assert_debug_snapshot!(recipe);
}

#[test]
fn test_conditionals_recipe_snapshot() {
    let source = load_test_recipe("conditionals.yaml");
    let recipe = parse_recipe_from_source(&source).expect("Failed to parse conditionals recipe");
    insta::assert_debug_snapshot!(recipe);
}

#[test]
fn test_complex_recipe_snapshot() {
    let source = load_test_recipe("complex.yaml");
    let recipe = parse_recipe_from_source(&source).expect("Failed to parse complex recipe");
    insta::assert_debug_snapshot!(recipe);
}

#[test]
fn test_run_exports_recipe_snapshot() {
    let source = load_test_recipe("run_exports.yaml");
    let recipe = parse_recipe_from_source(&source).expect("Failed to parse run_exports recipe");
    insta::assert_debug_snapshot!(recipe);
}

#[test]
fn test_templates_extract_variables() {
    let source = load_test_recipe("templates.yaml");
    let recipe = parse_recipe_from_source(&source).expect("Failed to parse templates recipe");
    let vars = recipe.used_variables();
    insta::assert_debug_snapshot!(vars);
}

#[test]
fn test_conditionals_extract_variables() {
    let source = load_test_recipe("conditionals.yaml");
    let recipe = parse_recipe_from_source(&source).expect("Failed to parse conditionals recipe");
    let vars = recipe.used_variables();
    insta::assert_debug_snapshot!(vars);
}

#[test]
fn test_complex_extract_variables() {
    let source = load_test_recipe("complex.yaml");
    let recipe = parse_recipe_from_source(&source).expect("Failed to parse complex recipe");
    let vars = recipe.used_variables();
    insta::assert_debug_snapshot!(vars);
}
