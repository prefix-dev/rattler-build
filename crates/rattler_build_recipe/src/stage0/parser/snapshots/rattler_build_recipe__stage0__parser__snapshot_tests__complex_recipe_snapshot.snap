---
source: crates/rattler_build_recipe/src/stage0/parser/snapshot_tests.rs
expression: recipe
---
SingleOutputRecipe {
    schema_version: None,
    context: {},
    package: Package {
        name: Template {
            template: JinjaTemplate {
                source: "${{ name | lower }}",
                variables: [
                    "name",
                ],
            },
            span: Span {
                start: 17,
                end: 17,
                start_line: 2,
                start_column: 9,
                end_line: 2,
                end_column: 9,
            },
        },
        version: Template {
            template: JinjaTemplate {
                source: "${{ version }}",
                variables: [
                    "version",
                ],
            },
            span: Span {
                start: 50,
                end: 50,
                start_line: 3,
                start_column: 12,
                end_line: 3,
                end_column: 12,
            },
        },
    },
    build: Build {
        number: Concrete {
            value: 0,
            span: Span {
                start: 0,
                end: 0,
                start_line: 0,
                start_column: 0,
                end_line: 0,
                end_column: 0,
            },
        },
        string: None,
        script: ConditionalList(
            [],
        ),
        noarch: None,
        python: PythonBuild {
            entry_points: ConditionalList(
                [],
            ),
            skip_pyc_compilation: ConditionalList(
                [],
            ),
            use_python_app_entrypoint: false,
            version_independent: false,
            site_packages_path: None,
        },
        skip: ConditionalList(
            [],
        ),
        always_copy_files: ConditionalList(
            [],
        ),
        always_include_files: ConditionalList(
            [],
        ),
        merge_build_and_host_envs: false,
        files: List(
            ConditionalList(
                [],
            ),
        ),
        dynamic_linking: DynamicLinking {
            rpaths: ConditionalList(
                [],
            ),
            binary_relocation: Boolean(
                Concrete {
                    value: true,
                    span: Span {
                        start: 0,
                        end: 0,
                        start_line: 0,
                        start_column: 0,
                        end_line: 0,
                        end_column: 0,
                    },
                },
            ),
            missing_dso_allowlist: ConditionalList(
                [],
            ),
            rpath_allowlist: ConditionalList(
                [],
            ),
            overdepending_behavior: None,
            overlinking_behavior: None,
        },
        variant: VariantKeyUsage {
            use_keys: ConditionalList(
                [],
            ),
            ignore_keys: ConditionalList(
                [],
            ),
            down_prioritize_variant: None,
        },
        prefix_detection: PrefixDetection {
            force_file_type: ForceFileType {
                text: ConditionalList(
                    [],
                ),
                binary: ConditionalList(
                    [],
                ),
            },
            ignore: Boolean(
                Concrete {
                    value: false,
                    span: Span {
                        start: 0,
                        end: 0,
                        start_line: 0,
                        start_column: 0,
                        end_line: 0,
                        end_column: 0,
                    },
                },
            ),
            ignore_binary_files: false,
        },
        post_process: [],
    },
    requirements: Requirements {
        build: ConditionalList(
            [
                Value(Template { template: JinjaTemplate { source: "${{ compiler(\"c\") }}", variables: ["compiler"] }, span: Span { start: 482, end: 482, start_line: 20, start_column: 7, end_line: 20, end_column: 7 } }),
                Value(Template { template: JinjaTemplate { source: "${{ compiler(\"cxx\") }}", variables: ["compiler"] }, span: Span { start: 511, end: 511, start_line: 21, start_column: 7, end_line: 21, end_column: 7 } }),
                Conditional(Conditional { condition: unix, then: ListOrItem(Parsed(MatchSpec { name: Some(PackageName { normalized: None, source: "make" }), version: None, build: None, build_number: None, file_name: None, extras: None, channel: None, subdir: None, namespace: None, md5: None, sha256: None, url: None, license: None })), else: ListOrItem([]) }),
                Conditional(Conditional { condition: win, then: ListOrItem(Parsed(MatchSpec { name: Some(PackageName { normalized: None, source: "cmake" }), version: None, build: None, build_number: None, file_name: None, extras: None, channel: None, subdir: None, namespace: None, md5: None, sha256: None, url: None, license: None })), else: ListOrItem([]) }),
            ],
        ),
        host: ConditionalList(
            [
                Value(Template { template: JinjaTemplate { source: "python ${{ python }}", variables: ["python"] }, span: Span { start: 614, end: 614, start_line: 27, start_column: 7, end_line: 27, end_column: 7 } }),
                Value(Concrete { value: Parsed(MatchSpec { name: Some(PackageName { normalized: None, source: "setuptools" }), version: Some(Range(GreaterEquals, Version { version: [[0], [45]], local: [] })), build: None, build_number: None, file_name: None, extras: None, channel: None, subdir: None, namespace: None, md5: None, sha256: None, url: None, license: None }), span: Span { start: 641, end: 641, start_line: 28, start_column: 7, end_line: 28, end_column: 7 } }),
                Value(Concrete { value: Parsed(MatchSpec { name: Some(PackageName { normalized: None, source: "wheel" }), version: None, build: None, build_number: None, file_name: None, extras: None, channel: None, subdir: None, namespace: None, md5: None, sha256: None, url: None, license: None }), span: Span { start: 663, end: 663, start_line: 29, start_column: 7, end_line: 29, end_column: 7 } }),
                Conditional(Conditional { condition: build_type == "cuda", then: ListOrItem(Deferred("cuda-toolkit ${{ cuda_version }}")), else: ListOrItem([]) }),
            ],
        ),
        run: ConditionalList(
            [
                Value(Template { template: JinjaTemplate { source: "python ${{ python }}", variables: ["python"] }, span: Span { start: 758, end: 758, start_line: 33, start_column: 7, end_line: 33, end_column: 7 } }),
                Value(Concrete { value: Parsed(MatchSpec { name: Some(PackageName { normalized: None, source: "numpy" }), version: Some(Group(And, [Range(GreaterEquals, Version { version: [[0], [1], [19]], local: [] }), Range(Less, Version { version: [[0], [2], [0, 'a', 0]], local: [] })])), build: None, build_number: None, file_name: None, extras: None, channel: None, subdir: None, namespace: None, md5: None, sha256: None, url: None, license: None }), span: Span { start: 785, end: 785, start_line: 34, start_column: 7, end_line: 34, end_column: 7 } }),
                Value(Template { template: JinjaTemplate { source: "${{ pin_compatible(\"numpy\") }}", variables: ["pin_compatible"] }, span: Span { start: 811, end: 811, start_line: 35, start_column: 7, end_line: 35, end_column: 7 } }),
                Conditional(Conditional { condition: linux, then: ListOrItem(Parsed(MatchSpec { name: Some(PackageName { normalized: None, source: "libgcc-ng" }), version: Some(Range(GreaterEquals, Version { version: [[0], [9], [3], [0]], local: [] })), build: None, build_number: None, file_name: None, extras: None, channel: None, subdir: None, namespace: None, md5: None, sha256: None, url: None, license: None })), else: ListOrItem([]) }),
                Conditional(Conditional { condition: osx, then: ListOrItem(Parsed(MatchSpec { name: Some(PackageName { normalized: None, source: "libcxx" }), version: Some(Range(GreaterEquals, Version { version: [[0], [10], [0], [0]], local: [] })), build: None, build_number: None, file_name: None, extras: None, channel: None, subdir: None, namespace: None, md5: None, sha256: None, url: None, license: None })), else: ListOrItem([]) }),
                Conditional(Conditional { condition: build_type == "cuda", then: ListOrItem(Deferred("cuda-toolkit ${{ cuda_version }}")), else: ListOrItem([]) }),
            ],
        ),
        run_constraints: ConditionalList(
            [
                Value(Concrete { value: Parsed(MatchSpec { name: Some(PackageName { normalized: None, source: "scipy" }), version: Some(Group(And, [Range(GreaterEquals, Version { version: [[0], [1], [5]], local: [] }), Range(Less, Version { version: [[0], [2], [0, 'a', 0]], local: [] })])), build: None, build_number: None, file_name: None, extras: None, channel: None, subdir: None, namespace: None, md5: None, sha256: None, url: None, license: None }), span: Span { start: 1033, end: 1033, start_line: 43, start_column: 7, end_line: 43, end_column: 7 } }),
                Conditional(Conditional { condition: extra_deps, then: ListOrItem(Parsed(MatchSpec { name: Some(PackageName { normalized: None, source: "pandas" }), version: Some(Range(GreaterEquals, Version { version: [[0], [1], [0]], local: [] })), build: None, build_number: None, file_name: None, extras: None, channel: None, subdir: None, namespace: None, md5: None, sha256: None, url: None, license: None })), else: ListOrItem([]) }),
            ],
        ),
        run_exports: RunExports {
            noarch: ConditionalList(
                [],
            ),
            strong: ConditionalList(
                [
                    Value(Template { template: JinjaTemplate { source: "${{ pin_subpackage(name, max_pin=\"x.x\") }}", variables: ["name", "pin_subpackage"] }, span: Span { start: 1133, end: 1133, start_line: 48, start_column: 9, end_line: 48, end_column: 9 } }),
                ],
            ),
            strong_constraints: ConditionalList(
                [],
            ),
            weak: ConditionalList(
                [
                    Conditional(Conditional { condition: has_cpp_lib, then: ListOrItem(Parsed(MatchSpec { name: Some(PackageName { normalized: None, source: "libstdcxx-ng" }), version: Some(Range(GreaterEquals, Version { version: [[0], [9], [3], [0]], local: [] })), build: None, build_number: None, file_name: None, extras: None, channel: None, subdir: None, namespace: None, md5: None, sha256: None, url: None, license: None })), else: ListOrItem([]) }),
                ],
            ),
            weak_constraints: ConditionalList(
                [],
            ),
        },
        ignore_run_exports: IgnoreRunExports {
            by_name: ConditionalList(
                [
                    Value(Concrete { value: "qt", span: Span { start: 1290, end: 1290, start_line: 54, start_column: 9, end_line: 54, end_column: 9 } }),
                    Value(Concrete { value: "tk", span: Span { start: 1301, end: 1301, start_line: 55, start_column: 9, end_line: 55, end_column: 9 } }),
                ],
            ),
            from_package: ConditionalList(
                [
                    Value(Concrete { value: "python", span: Span { start: 1330, end: 1330, start_line: 57, start_column: 9, end_line: 57, end_column: 9 } }),
                ],
            ),
        },
    },
    about: About {
        homepage: Some(
            Template {
                template: JinjaTemplate {
                    source: "https://github.com/${{ org }}/${{ name }}",
                    variables: [
                        "name",
                        "org",
                    ],
                },
                span: Span {
                    start: 87,
                    end: 87,
                    start_line: 6,
                    start_column: 13,
                    end_line: 6,
                    end_column: 13,
                },
            },
        ),
        license: Some(
            Template {
                template: JinjaTemplate {
                    source: "${{ license }}",
                    variables: [
                        "license",
                    ],
                },
                span: Span {
                    start: 140,
                    end: 140,
                    start_line: 7,
                    start_column: 12,
                    end_line: 7,
                    end_column: 12,
                },
            },
        ),
        license_file: ConditionalList(
            [
                Value(Concrete { value: "LICENSE.txt", span: Span { start: 173, end: 173, start_line: 8, start_column: 17, end_line: 8, end_column: 17 } }),
            ],
        ),
        license_family: None,
        summary: Some(
            Template {
                template: JinjaTemplate {
                    source: "${{ summary }}",
                    variables: [
                        "summary",
                    ],
                },
                span: Span {
                    start: 196,
                    end: 196,
                    start_line: 9,
                    start_column: 12,
                    end_line: 9,
                    end_column: 12,
                },
            },
        ),
        description: Some(
            Template {
                template: JinjaTemplate {
                    source: "${{ description }}\n\nThis is a multi-line description that can contain\ntemplates and other formatting.\n",
                    variables: [
                        "description",
                    ],
                },
                span: Span {
                    start: 234,
                    end: 234,
                    start_line: 11,
                    start_column: 5,
                    end_line: 11,
                    end_column: 5,
                },
            },
        ),
        documentation: Some(
            Template {
                template: JinjaTemplate {
                    source: "https://${{ name }}.readthedocs.io",
                    variables: [
                        "name",
                    ],
                },
                span: Span {
                    start: 361,
                    end: 361,
                    start_line: 15,
                    start_column: 18,
                    end_line: 15,
                    end_column: 18,
                },
            },
        ),
        repository: Some(
            Template {
                template: JinjaTemplate {
                    source: "https://github.com/${{ org }}/${{ name }}",
                    variables: [
                        "name",
                        "org",
                    ],
                },
                span: Span {
                    start: 410,
                    end: 410,
                    start_line: 16,
                    start_column: 15,
                    end_line: 16,
                    end_column: 15,
                },
            },
        ),
    },
    extra: Extra {
        recipe_maintainers: ConditionalList(
            [
                Value(Template { template: JinjaTemplate { source: "${{ maintainer1 }}", variables: ["maintainer1"] }, span: Span { start: 1373, end: 1373, start_line: 61, start_column: 7, end_line: 61, end_column: 7 } }),
                Value(Template { template: JinjaTemplate { source: "${{ maintainer2 }}", variables: ["maintainer2"] }, span: Span { start: 1400, end: 1400, start_line: 62, start_column: 7, end_line: 62, end_column: 7 } }),
            ],
        ),
    },
    source: [],
    tests: [],
}
